src/relax/transform/extract_dataflow_blocks.cc

```diff
@@ -0,0 +1,140 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/relax/transform/extract_dataflow_blocks.cc
+ * \brief Pass for extracting groups of pure operations without
+ *   dataflow into dataflow blocks.
+ */
+
+#include <tvm/relax/expr.h>
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/transform.h>
+#include <tvm/relax/utils.h>
+
+namespace tvm {
+namespace relax {
+
+class DataflowBlockExtractor : public ExprMutator {
+ public:
+  explicit DataflowBlockExtractor(size_t min_size) : ExprMutator(), min_size_(min_size) {}
+
+  Expr VisitExpr_(const SeqExprNode* seq) override {
+    Array<BindingBlock> new_blocks;
+    Expr new_body = VisitExpr(seq->body);
+    bool changed = !new_body.same_as(seq->body);
+    for (auto block : seq->blocks) {
+      BindingBlock new_block = this->VisitBindingBlock(block);
+      changed = changed || !new_block.same_as(block);
+      if (new_block.as<DataflowBlock>()) {
+        new_blocks.push_back(new_block);
+        continue;
+      }
+
+      // for a binding block, attempt to extract dataflow blocks inside
+      auto binding_block = Downcast<BindingBlock>(new_block);
+      bool dataflow_streak = false;
+      Array<Binding> dataflow_bindings;
+      Array<Binding> non_dataflow_bindings;
+      for (size_t i = 0; i < binding_block->bindings.size(); i++) {
+        auto binding = binding_block->bindings[i];
+        Expr value = GetBoundValue(binding);
+        // dataflow values: not an if node and not an impure call
+        bool is_dataflow =
+            (!value.as<IfNode>()) && (!(value.as<CallNode>() && IsImpureCall(Downcast<Call>(value))));
+        if (!dataflow_streak) {
+          // we can start a dataflow streak
+          if (is_dataflow) {
+            dataflow_streak = true;
+            dataflow_bindings = {binding};
+          } else {
+            non_dataflow_bindings.push_back(binding);
+          }
+        } else {
+          if (is_dataflow) {
+            // extend the streak
+            dataflow_bindings.push_back(binding);
+          } else {
+            // this is the end of the streak
+            dataflow_streak = false;
+
+            // if the df block is below the minimum length, combine the blocks
+            // and reset the dataflow collection
+            if (dataflow_bindings.size() < min_size_) {
+              non_dataflow_bindings.insert(non_dataflow_bindings.end(), dataflow_bindings.begin(),
```

Comment:
What happens if `dataflow_bindings.size() < min_size_`, but the preceding block was also a dataflow block?  As a user, I'd want it to be merged into the preceding block anyways, but it looks like it would be left as-is.

Reply:
That's a good point. Currently it's merged into the following (non-dataflow) block. This is a case worth tracking, I'll think about how to get it.